pennylane/compiler/python_compiler/transforms/diagonalize_measurements.py

```diff
@@ -0,0 +1,108 @@
+    (for example, qml.H(0).diagonalizing_gates())
+  * Unlike the current tape-based implementation of the transform, it doesn't allow for diagonalization of a subset
+    of observables.
+  * Unlike the current tape-based implementation of the transform, conversion to measurements based on eigvals
+  and wires (rather than the PauliZ observable) is not currently supported.
+"""
+
+from dataclasses import dataclass
+
+from xdsl import context, passes, pattern_rewriter
+from xdsl.dialects import builtin
+from xdsl.rewriter import InsertPoint
+
+from pennylane import ops as pl_ops
+
+from ..quantum_dialect import CustomOp, NamedObservable, NamedObservableAttr, NamedObsOp
+from .api import compiler_transform
+
+
+def _diagonalize(obs: NamedObsOp) -> bool:
+    """Whether to diagonalize a given observable."""
+    if obs.type.data in ("PauliZ", "Identity"):
+        return False
+    if obs.type.data in ("PauliX", "PauliY"):
+        return True
+    raise NotImplementedError(f"Observable {obs.type.data} is not supported for diagonalization")
+
+
+def _diagonalizing_gates(obs: NamedObsOp):
+    """Get the names of the observable's diagonalizing gates from the corresponding
+    PennyLane Operation, and use them to get an xDSL representation of the diagonalizing
+    gates.
+
+    Returns a list of CustomOps representing the chain of diagonalizing gates, and the
+    new SSA value for the observable qubit wire following diagonalization."""
+    pl_obs = getattr(pl_ops, obs.type.data)(0)
+    gates = [o.name for o in pl_obs.diagonalizing_gates()]
+
+    diagonalizing_gates = []
+    qubit = obs.qubit
+
+    for name in gates:
+        diagonalizing_gates.append(CustomOp(in_qubits=qubit, gate_name=name))
+        qubit = diagonalizing_gates[-1].out_qubits[0]
+
+    return diagonalizing_gates, qubit
+
+
+class DiagonalizeFinalMeasurementsPattern(
+    pattern_rewriter.RewritePattern
+):  # pylint: disable=too-few-public-methods
+    """RewritePattern for diagonalizing final measurements."""
+
+    # pylint: disable=no-self-use
+    @pattern_rewriter.op_type_rewrite_pattern
+    def match_and_rewrite(
+        self, observable: NamedObsOp, rewriter: pattern_rewriter.PatternRewriter
+    ):  # pylint: disable=arguments-differ
+        """Replace non-diagonalized observables with their diagonalizing gates and PauliZ."""
+
+        if _diagonalize(observable):
+
+            diagonalizing_gates, obs_qubit = _diagonalizing_gates(observable)
+            for gate in diagonalizing_gates:
+                rewriter.insert_op(gate, InsertPoint.before(observable))
+
+            diag_obs = NamedObsOp(
+                qubit=obs_qubit, obs_type=NamedObservableAttr(NamedObservable("PauliZ"))
+            )
+            rewriter.replace_op(observable, diag_obs)
+
+
+@dataclass(frozen=True)
+class DiagonalizeFinalMeasurementsPass(passes.ModulePass):
+    """Pass for merging consecutive composable rotation gates."""
+
+    name = "diagonalize-measurements"
```

Comment:
Would it be more accurate to use ``diagonalize-final-measurements``?

Reply:
That's actually what I had initially, because yes, its clearer now that we also think about diagonalizing MCMs - but this mirrors the PL transform name. So I'm on the fence 🤷‍♀️

Reply:
I changed it to "diagonalize final measurements" for clarity between that and "diagonalize mcms".